gradle-plugin/src/main/groovy/io/realm/gradle/RealmTransformer.groovy

```diff
@@ -0,0 +1,130 @@
+import com.google.common.collect.Sets
+import groovy.io.FileType
+import io.realm.annotations.Ignore
+import javassist.ClassPool
+import javassist.LoaderClassPath
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+import static com.android.build.api.transform.QualifiedContent.*
+
+class RealmTransformer extends Transform {
+
+    private Logger logger = LoggerFactory.getLogger('realm-logger')
+
+    @Override
+    String getName() {
+        return "RealmTransformer"
+    }
+
+    @Override
+    Set<ContentType> getInputTypes() {
+        return ImmutableSet.<ContentType> of(DefaultContentType.CLASSES)
+    }
+
+    @Override
+    Set<Scope> getScopes() {
+        return Sets.immutableEnumSet(
+                Scope.PROJECT)
+    }
+
+    @Override
+    boolean isIncremental() {
+        return true
+    }
+
+    @Override
+    void transform(Context context, Collection<TransformInput> inputs, Collection<TransformInput> referencedInputs,
+                   TransformOutputProvider outputProvider, boolean isIncremental)
+            throws IOException, TransformException, InterruptedException {
+
+        final ArrayList<File> folders = []
+        inputs.each { TransformInput input ->
+            logger.info "Directory inputs: ${input.directoryInputs*.file}"
+            folders.addAll(input.directoryInputs*.file)
+        }
+
+        referencedInputs.each { TransformInput input ->
+            folders.addAll(input.directoryInputs*.file)
+            folders.addAll(input.jarInputs*.file)
+        }
+
+
+        BiMap<File, String> classFiles = HashBiMap.create()
+
+        folders.each { File folder ->
+            folder.eachFileRecurse(FileType.FILES) { File file ->
+                if (file.name.endsWith('.class')) {
+                    classFiles.put(
+                            file,
+                            file.canonicalPath
+                                    .replace(folder.canonicalPath, '')
+                                    .replaceFirst('/', '')
+                                    .replace('/', '.')
+                                    .replace('.class', ''))
+                }
+            }
+        }
+
+        // Don't use ClassPool.getDefault(). Doing so consecutive builds in the same run (e.g. debug+release)
+        // will use a cached object and all the classes will result frozen.
+        ClassPool classPool = new ClassPool(null)
+        classPool.appendSystemPath()
+        classPool.appendClassPath(new LoaderClassPath(getClass().getClassLoader()))
+
+        folders.each { File folder ->
+            classPool.appendClassPath(folder.canonicalPath)
+        }
+
+        logger.info "Contains io.realm.RealmList: ${classPool.getOrNull('io.realm.RealmList')}"
+
+        def proxyClasses = classFiles.findAll { key, value -> key.name.endsWith('RealmProxy.class') }
+        logger.info "Proxy Classes: ${proxyClasses*.value}"
+
+        def modelClasses = proxyClasses.collect { key, value ->
```

Comment:
This was a major headache to solve. The problem is that `RealmObject` is not in the Javassist pool since the scope is "Scope.PROJECT" only. This means we cannot use `RealmObject` to identify model classes.